### PR TITLE
cephadm: fix `cephadm shell --name <daemon-name>` for stopped/failed daemon

### PR DESCRIPTION
**Pull request by adk3798**
This previously would always try to use 'podman
inspect' on the running container of the daemon,
but this doesn't work if the daemon is stopped
or failed. Doing this for stopped/failed daemons
is a valid use case as we recommend cephadm shell
with --name for running debugging tools (often
for OSDs)

Fixes: https://tracker.ceph.com/issues/64879

Signed-off-by: Adam King <adking@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/cephadm/cephadm.py [hunk 1]
```diff
@@ -483,14 +483,51 @@ def daemon_name_or_type(daemon: Dict[str, str]) -> str:
     if by_name and '.' not in daemon_filter:
         logger.warning(f'Trying to get container info using invalid daemon name {daemon_filter}')
         return None
-    daemons = list_daemons(ctx, detail=False)
-    matching_daemons = [d for d in daemons if daemon_name_or_type(d) == daemon_filter and d['fsid'] == ctx.fsid]
+    if by_name:
+        matching_daemons = list_daemons(ctx, daemon_name=daemon_filter)
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:487]: IMO, this block might be best in a helper function.

File: src/cephadm/cephadm.py [hunk 2]
```diff
@@ -483,14 +483,51 @@ def daemon_name_or_type(daemon: Dict[str, str]) -> str:
...
+        if len(matching_daemons) > 1:
+            logger.error(f'Found multiple daemons sharing same name: {daemon_filter}')
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:489]: If this is non-fatal maybe this should be a warning instead.

File: src/cephadm/cephadm.py [hunk 3]
```diff
@@ -483,14 +483,51 @@ def daemon_name_or_type(daemon: Dict[str, str]) -> str:
...
+            matched_daemon = None
+            for d in matching_daemons:
+                if 'state' in d and d['state'] == 'running':
+                    matched_daemon = d
+                    break
+            if not matched_daemon:
+                matched_daemon = matching_daemons[0]
+            matching_daemons = [matched_daemon]
+    else:
+        daemons = list_daemons(ctx, detail=False)
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:501]: In the other branch detail is not set to anything. Are they intended to be different?

File: src/cephadm/cephadm.py [hunk 4]
```diff
@@ -483,14 +483,51 @@ def daemon_name_or_type(daemon: Dict[str, str]) -> str:
...
+            and 'container_image_name' in matching_daemons[0]
+            and matching_daemons[0]['container_image_name']
+        ):
+            # this daemon contianer is not running so the regular `podman/docker inspect` on the
+            # container will not help us. If we have the image name from the list_daemons output
+            # we can try that.
+            image_name = matching_daemons[0]['container_image_name']
+            out, _, code = get_container_stats_by_image_name(ctx, ctx.container_engine.path, image_name)
+            if not code:
+                # keep in mind, the daemon contianer is not running, so no container id here
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:517]: sp: `contianer`

File: src/cephadm/cephadm.py [hunk 5]
```diff
@@ -3671,6 +3709,18 @@ def get_container_stats(ctx: CephadmContext, container_path: str, fsid: str, dae
             break
     return out, err, code
 
+
+def get_container_stats_by_image_name(ctx: CephadmContext, container_path: str, image_name: str) -> Tuple[str, str, int]:
+    # returns image id, created time, and ceph version if available
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:3714]: make it a docstring, not a comment.

File: src/cephadm/cephadm.py [hunk 6]
```diff
@@ -3656,6 +3693,7 @@ def get_daemon_description(ctx, fsid, name, detail=False, legacy_dir=None):
 
 
 def get_container_stats(ctx: CephadmContext, container_path: str, fsid: str, daemon_type: str, daemon_id: str) -> Tuple[str, str, int]:
+    # returns container id, image name, image id, created time, and ceph version if available
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:3696]: make it a docstring, not a comment.

File: src/cephadm/cephadm.py [hunk 7]
```diff
@@ -3452,6 +3457,8 @@ def list_daemons(ctx, detail=True, legacy_dir=None):
                 for j in os.listdir(os.path.join(data_dir, i)):
                     if '.' in j and os.path.isdir(os.path.join(data_dir, fsid, j)):
                         name = j
+                        if daemon_name and name != daemon_name:
```
Comment (phlogistonjohn) [inline src/cephadm/cephadm.py:3497]: This is fine for now, but it seems to me the whole list_daemons thing is crying out for a nice refactoring ;-)

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]

File: src/cephadm/cephadm.py (hunk 3 above)
Comment (adk3798) [reply src/cephadm/cephadm.py:501]: yes, `detail=False` excludes the container image field from the result. in the case where the container is running, that is okay, but when it isn't we need that field.

Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:501]: OK, can you make that comment in the code, please?

**Review by phlogistonjohn — APPROVED** [review]
looks ok to me.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-03-16_05:03:17-orch:cephadm-wip-adk-testing-2024-03-15-1513-distro-default-smithi/

mostly cluster log stuff + mds_upgrade_sequence. Other failures were either known or not reproducible.

**Pull request merged**